### PR TITLE
MM  418  - Add --exclude option to subscriptions command #418

### PR DESCRIPTION
**Pull request by sibasankarnayak**
Added `--exclude` command for disabling the notification for those mentioned repos

Ticket Link 
https://github.com/mattermost/mattermost-plugin-github/issues/418

### CODE REVIEW

File: server/plugin/command.go [hunk 1]
```diff
@@ -267,25 +267,36 @@ func (p *Plugin) handleSubscriptionsList(_ *plugin.Context, args *model.CommandA
...
+	var isExcludeRepo bool
+	var excludeRepo string
 	if len(parameters) > 1 {
 		var optionList []string
-
 		for _, element := range parameters[1:] {
-			if isFlag(element) {
+			switch {
+			case parseFlag(element) == "exclude":
+				isExcludeRepo = true
```
Comment (hanzei) [inline server/plugin/command.go:291]: Could this make use of `SubscriptionFlags`?

File: server/plugin/command.go [hunk 2]
```diff
@@ -267,25 +267,36 @@ func (p *Plugin) handleSubscriptionsList(_ *plugin.Context, args *model.CommandA
...
+			case isFlag(element):
 				flags.AddFlag(parseFlag(element))
-			} else {
+			case isExcludeRepo && excludeRepo == "":
+				excludeRepo = element
+			default:
 				optionList = append(optionList, element)
 			}
 		}
-
```
Comment (hanzei) [inline server/plugin/command.go:288]: Unrelated change

File: server/plugin/command.go [hunk 3]
```diff
@@ -267,25 +267,36 @@ func (p *Plugin) handleSubscriptionsList(_ *plugin.Context, args *model.CommandA
...
+		txt += fmt.Sprintf("* `%s` - %s", strings.Trim(repo, "/"), "notification : disabled")
+		txt += "\n"
+	}
 	return txt
 }
 
 func (p *Plugin) handleSubscribesAdd(_ *plugin.Context, args *model.CommandArgs, parameters []string, userInfo *GitHubUserInfo) string {
 	features := "pulls,issues,creates,deletes"
 	flags := SubscriptionFlags{}
-
```
Comment (hanzei) [inline server/plugin/command.go:277]: Unrelated change

File: server/plugin/command.go [hunk 4]
```diff
@@ -313,10 +324,30 @@ func (p *Plugin) handleSubscribesAdd(_ *plugin.Context, args *model.CommandArgs,
 		if err := p.SubscribeOrg(ctx, githubClient, args.UserId, owner, args.ChannelId, features, flags); err != nil {
 			return err.Error()
 		}
-
-		return fmt.Sprintf("Successfully subscribed to organization %s.", owner)
+		var subOrgMsg = fmt.Sprintf("Successfully subscribed to organization %s.", owner)
+		if isExcludeRepo {
+			var excludeMsg string
+			for _, val := range strings.Split(excludeRepo, ",") {
+				notificationOffRepoOwner, NotificationOffRepo := parseOwnerAndRepo(val, p.getBaseURL())
```
Comment (hanzei) [inline server/plugin/command.go:331]: Should empty spaces around `val` be allowed?

File: server/plugin/command.go [hunk 5]
```diff
@@ -313,10 +324,30 @@ func (p *Plugin) handleSubscribesAdd(_ *plugin.Context, args *model.CommandArgs,
...
+				if notificationOffRepoOwner != owner {
+					return fmt.Sprintf("--exclude repository  %s is not of subscribed organization .", NotificationOffRepo)
+				}
```
Comment (hanzei) [inline server/plugin/command.go:334]: Should the exclude flag accept a list of repos instead of a list of org/repo?

File: server/plugin/command.go [hunk 6]
```diff
@@ -339,7 +370,9 @@ func (p *Plugin) handleUnsubscribe(_ *plugin.Context, args *model.CommandArgs, p
 	}
 
 	repo := parameters[0]
-
+	if err := p.EnableNotificationTurnedOffRepo(repo); err != nil {
+		return "Encountered an error trying to remove from notify disabled list. Please try again."
```
Comment (hanzei) [inline server/plugin/command.go:374]: Why is `err` to getting wrapped?

File: server/plugin/subscriptions.go [hunk 7]
```diff
@@ -146,22 +148,31 @@ func (p *Plugin) Subscribe(ctx context.Context, githubClient *github.Client, use
...
-
 	return p.Subscribe(ctx, githubClient, userID, org, "", channelID, features, flags)
 }
 
+func (p *Plugin) CheckIsNotificationOff(repoName string) (bool, error) {
+	subscription, err := p.GetNotificationTurnedOffRepo()
+	if err != nil {
+		return false, err
+	}
+	p.API.LogWarn("subscription for notification off ", "subscription", subscription)
```
Comment (hanzei) [inline server/plugin/subscriptions.go:169]: Debugging?

File: server/plugin/utils.go [hunk 8]
```diff
@@ -332,3 +333,19 @@ func getToDoDisplayText(baseURL, title, url, notifType string) string {
 
 	return fmt.Sprintf("* %s %s %s\n", repoPart, notifType, titlePart)
 }
+
+// ItemExists check if the element exist in the passed array
+func ItemExists(arrayType interface{}, item interface{}) (bool, int) {
```
Comment (hanzei) [inline server/plugin/utils.go:338]: This can accept a string and a slice of strings. No need for fancy reflection.

File: server/plugin/webhook.go [hunk 9]
```diff
@@ -92,47 +92,79 @@ func (p *Plugin) handleWebhook(w http.ResponseWriter, r *http.Request) {
 	switch event := event.(type) {
 	case *github.PullRequestEvent:
 		repo = event.GetRepo()
+		var shouldNotNotify, err = p.CheckIsNotificationOff(*repo.FullName)
+		if err != nil || shouldNotNotify {
+			return
+		}
```
Comment (hanzei) [inline server/plugin/webhook.go:98]: Should the error get logged?

File: server/plugin/webhook.go [hunk 10]
```diff
@@ -92,47 +92,79 @@ func (p *Plugin) handleWebhook(w http.ResponseWriter, r *http.Request) {
 	switch event := event.(type) {
 	case *github.PullRequestEvent:
 		repo = event.GetRepo()
+		var shouldNotNotify, err = p.CheckIsNotificationOff(*repo.FullName)
```
Comment (hanzei) [inline server/plugin/webhook.go:95]: Are there cases where `CheckIsNotificationOff` should not get checked? If not, can the check the unified?

**Review by hanzei — CHANGES_REQUESTED** [review]
Thanks for PR :+1:

**Comment by hanzei** [comment]
@sibasankar-demansol Would you please remove the unrelated white space removals from the PR? That would make it much easier to review.

File: server/plugin/command.go (hunk 2 above)
Comment (sibasankar-demansol) [reply server/plugin/command.go:288]: resolved

File: server/plugin/command.go (hunk 3 above)
Comment (sibasankar-demansol) [reply server/plugin/command.go:277]: resolved

File: server/plugin/command.go (hunk 4 above)
Comment (sibasankar-demansol) [reply server/plugin/command.go:331]: yes, added trim extra spaces from the name

File: server/plugin/command.go (hunk 5 above)
Comment (sibasankar-demansol) [reply server/plugin/command.go:334]: currently it is taking org/repo ,
do we want to take that feature ?

File: server/plugin/command.go (hunk 6 above)
Comment (sibasankar-demansol) [reply server/plugin/command.go:374]: added a logWarn for the error if occurred

File: server/plugin/utils.go (hunk 8 above)
Comment (sibasankar-demansol) [reply server/plugin/utils.go:338]: resolved , made it simple

File: server/plugin/webhook.go (hunk 9 above)
Comment (sibasankar-demansol) [reply server/plugin/webhook.go:98]: resolved , logged the error

File: server/plugin/webhook.go (hunk 10 above)
Comment (sibasankar-demansol) [reply server/plugin/webhook.go:95]: Sorry , i didn't got this 
currently for every event handled in webhook it is configured

File: server/plugin/command.go (hunk 4 above)
Comment (hanzei) [reply server/plugin/command.go:334]: Why isn't `TrimSpace` used?

File: server/plugin/command.go (hunk 6 above)
Comment (hanzei) [reply server/plugin/command.go:378]: I'm 1/5 that the code should handle errors in commands consistently which mean use wrapping. The error message might also be handy for the end user.

File: server/plugin/webhook.go (hunk 10 above)
Comment (hanzei) [reply server/plugin/webhook.go:95]: Right now the code checks `CheckIsNotificationOff` in every `case`. Can this be unified to only check once?

File: server/plugin/command.go (hunk 5 above)
Comment (hanzei) [reply server/plugin/command.go:337]: The feature seems crucial to me as subscriptions usually happen on an org level for this plugin.

**Review by hanzei — CHANGES_REQUESTED** [review]
There are a couple of outstanding things.

I also just wanted to check if you saw my comment in https://github.com/mattermost/mattermost-plugin-github/pull/457#issuecomment-880647377. There are still some unrelated changes.

File: server/plugin/subscriptions.go [hunk 11]
```diff
@@ -255,6 +265,66 @@ func (p *Plugin) StoreSubscriptions(s *Subscriptions) error {
 	return nil
 }
 
+func (p *Plugin) GetNotificationTurnedOffRepo() ([]string, error) {
```
Comment (jfrerich) [inline server/plugin/subscriptions.go:268]:
```suggestion
func (p *Plugin) GetExcludedNotificationRepos() ([]string, error) {
```
I think this name would be more intention revealing.

File: server/plugin/command.go [hunk 12]
```diff
@@ -267,21 +267,34 @@ func (p *Plugin) handleSubscriptionsList(_ *plugin.Context, args *model.CommandA
...
 	}
-
+	subscription, err := p.GetNotificationTurnedOffRepo()
+	if err != nil {
+		return err.Error()
+	}
+	for _, repo := range subscription {
+		txt += fmt.Sprintf("* `%s` - %s", strings.Trim(repo, "/"), "notification : disabled")
+		txt += "\n"
+	}
```
Comment (jfrerich) [inline server/plugin/command.go:277]:
```suggestion
	excludedRepos, err := p.GetNotificationTurnedOffRepo()
	if err != nil {
		return err.Error()
	}
	for _, repo := range excludedRepos {
		txt += fmt.Sprintf("* `%s` - %s", strings.Trim(repo, "/"), "notification : disabled")
		txt += "\n"
	}
```

File: server/plugin/command.go [hunk 13]
```diff
@@ -552,12 +589,23 @@ func getAutocompleteData(config *Configuration) *model.AutocompleteData {
...
 	if config.GitHubOrg != "" {
-		flags := []model.AutocompleteListItem{{
-			HelpText: "Events triggered by organization members will not be delivered (the organization config should be set, otherwise this flag has not effect)",
-			Hint:     "(optional)",
-			Item:     "--exclude-org-member",
-		}}
-		subscriptionsAdd.AddStaticListArgument("Currently supports --exclude-org-member", false, flags)
+		exclude := []model.AutocompleteListItem{
+			{
+				HelpText: "notification for these repo will be turned off",
```
Comment (jfrerich) [inline server/plugin/command.go:594]:
```suggestion
				HelpText: "Notifications for these repos will be turned off",
```

File: server/plugin/command.go [hunk 14]
```diff
@@ -552,12 +589,23 @@ func getAutocompleteData(config *Configuration) *model.AutocompleteData {
...
+				Hint:     "(optional)",
+				Item:     "--exclude",
+			},
+		}
+		subscriptionsAdd.AddStaticListArgument("Currently supports --exclude", true, exclude)
+		subscriptionsAdd.AddTextArgument("Owner/repo to subscribe to", "[owner/repo]", "")
+		flags := []model.AutocompleteListItem{
+			{
+				HelpText: "Events triggered by organization members will not be delivered (the organization config should be set, otherwise this flag has not effect)",
```
Comment (jfrerich) [inline server/plugin/command.go:603]:
```suggestion
				HelpText: "Events triggered by organization members will not be delivered (the organization config should be set, otherwise this flag has no effect)",
```

File: server/plugin/subscriptions.go [hunk 15]
```diff
@@ -13,8 +13,10 @@ import (
 )
 
 const (
-	SubscriptionsKey     = "subscriptions"
-	excludeOrgMemberFlag = "exclude-org-member"
+	SubscriptionsKey = "subscriptions"
+	// SubscriptionsTunerOffNotification = "subscriptions-turned-off-notifications"
```
Comment (jfrerich) [inline server/plugin/subscriptions.go:17]: remove commented line.

File: server/plugin/subscriptions.go [hunk 16]
```diff
@@ -255,6 +265,66 @@ func (p *Plugin) StoreSubscriptions(s *Subscriptions) error {
...
+		return []string{}, nil
+	}
+	err := json.NewDecoder(bytes.NewReader(value)).Decode(&subscriptions)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not properly decode subscriptions key")
+	}
+	return subscriptions, nil
+}
+
+func (p *Plugin) StoreNotificationTurnedOffRepo(s string) error {
```
Comment (jfrerich) [inline server/plugin/subscriptions.go:284]:
```suggestion
func (p *Plugin) StoreExcludedNotificationRepo(s string) error {
```

File: server/plugin/subscriptions.go [hunk 17]
```diff
@@ -255,6 +265,66 @@ func (p *Plugin) StoreSubscriptions(s *Subscriptions) error {
...
+
+	return nil
+}
+func (p *Plugin) EnableNotificationTurnedOffRepo(s string) error {
+	var repoNames, err = p.GetNotificationTurnedOffRepo()
+	if err != nil {
+		return errors.Wrap(err, "error while getting previous value of key")
+	}
+	if len(repoNames) > 0 {
+		isDer, index := ItemExists(repoNames, s)
```
Comment (jfrerich) [inline server/plugin/subscriptions.go:312]: what is `isDer`? :) Perhaps just `exists` would work?

File: server/plugin/subscriptions.go [hunk 18]
```diff
@@ -146,22 +148,31 @@ func (p *Plugin) Subscribe(ctx context.Context, githubClient *github.Client, use
...
 func (p *Plugin) SubscribeOrg(ctx context.Context, githubClient *github.Client, userID, org, channelID, features string, flags SubscriptionFlags) error {
 	if org == "" {
 		return errors.New("invalid organization")
 	}
-
 	return p.Subscribe(ctx, githubClient, userID, org, "", channelID, features, flags)
 }
 
+func (p *Plugin) CheckIsNotificationOff(repoName string) (bool, error) {
+	subscription, err := p.GetNotificationTurnedOffRepo()
```
Comment (jfrerich) [inline server/plugin/subscriptions.go:165]: I believe `subscription` is actually a list of repos.
```suggestion
	repos, err := p.GetNotificationTurnedOffRepo()
```

File: server/plugin/webhook.go [hunk 19]
```diff
@@ -92,47 +92,135 @@ func (p *Plugin) handleWebhook(w http.ResponseWriter, r *http.Request) {
...
+		if err != nil || shouldNotNotify {
+			if err != nil {
+				p.API.LogWarn("Failed to check the check disabled notification repo list ", "error", err.Error())
```
Comment (jfrerich) [inline server/plugin/webhook.go:98]:
```suggestion
				p.API.LogWarn("Failed to check the disabled notification repo list ", "error", err.Error())
```

**Review by jfrerich — CHANGES_REQUESTED** [review]
I have a couple of suggestions and would like to see the other comments addressed before re-reviewing.  Thanks for the work!

File: server/plugin/webhook.go (hunk 19 above)
Comment (sibasankar-demansol) [reply server/plugin/webhook.go:98]: Done

File: server/plugin/subscriptions.go (hunk 18 above)
Comment (sibasankar-demansol) [reply server/plugin/subscriptions.go:165]: done

File: server/plugin/subscriptions.go (hunk 17 above)
Comment (sibasankar-demansol) [reply server/plugin/subscriptions.go:312]: done

File: server/plugin/subscriptions.go (hunk 16 above)
Comment (sibasankar-demansol) [reply server/plugin/subscriptions.go:284]: done

File: server/plugin/subscriptions.go (hunk 15 above)
Comment (sibasankar-demansol) [reply server/plugin/subscriptions.go:17]: Done

File: server/plugin/command.go (hunk 14 above)
Comment (sibasankar-demansol) [reply server/plugin/command.go:603]: Done

File: server/plugin/subscriptions.go (hunk 11 above)
Comment (sibasankar-demansol) [reply server/plugin/subscriptions.go:268]: Done

File: server/plugin/command.go (hunk 12 above)
Comment (sibasankar-demansol) [reply server/plugin/command.go:277]: done

File: server/plugin/command.go (hunk 13 above)
Comment (sibasankar-demansol) [reply server/plugin/command.go:594]: done

**Review by hanzei — CHANGES_REQUESTED** [review]
Blocking on my outstanding comments

File: server/plugin/webhook.go (hunk 10 above)
Comment (sibasankar-demansol) [reply server/plugin/webhook.go:95]: seems the event is a interface , so need a type switch to get the repository name from event so added in the existing switch only , let me know if we can do it in better way

File: server/plugin/command.go (hunk 5 above)
Comment (sibasankar-demansol) [reply server/plugin/command.go:337]: subscriptions can be done on both org and individual repos , but excluding repos can only be done it we are doing in org level.

so do we want to add this feature in repo level too ?

File: server/plugin/command.go (hunk 1 above)
Comment (sibasankar-demansol) [reply server/plugin/command.go:292]: Sure. , have made the changes

File: server/plugin/command.go (hunk 4 above)
Comment (sibasankar-demansol) [reply server/plugin/command.go:334]: have added this .

File: server/plugin/subscriptions.go (hunk 7 above)
Comment (sibasankar-demansol) [reply server/plugin/subscriptions.go]: yes , removed it

File: server/plugin/command.go (hunk 6 above)
Comment (sibasankar-demansol) [reply server/plugin/command.go:378]: should i remove this error check. 
we have already an previous error handling for unsubscribe part .

File: server/plugin/webhook.go [hunk 20]
```diff
@@ -92,47 +92,135 @@ func (p *Plugin) handleWebhook(w http.ResponseWriter, r *http.Request) {
 	switch event := event.(type) {
 	case *github.PullRequestEvent:
 		repo = event.GetRepo()
+		var shouldNotNotify, err = p.CheckIsNotificationOff(*repo.FullName)
+		if err != nil || shouldNotNotify {
+			if err != nil {
+				p.API.LogWarn("ailed to check the disabled notification repo list.", "error", err.Error())
```
Comment (jfrerich) [inline server/plugin/webhook.go:98]: The `F` in `Failed` was removed.

**Review by jfrerich — CHANGES_REQUESTED** [review]

File: server/plugin/webhook.go (hunk 20 above)
Comment (sibasankar-demansol) [reply server/plugin/webhook.go]: resolved

File: server/plugin/webhook.go (hunk 10 above)
Comment (jfrerich) [reply server/plugin/webhook.go:95]: Can you put the error checking into the `CheckIsNotificationOff` function and remove the redundant code?  Maybe rename the function also..  Each `case` could just have something like the following:

```go
	case *github.PullRequestEvent:
		repo = event.GetRepo()
		if p.IsNotificationOff(*repo.Fullname) {
        	return
		}
		handler = func() {
		  // do handling
		}
```

Comment (sibasankar-demansol) [reply server/plugin/webhook.go]: done

**Comment by maisnamrajusingh** [comment]
@hanzei  @jfrerich are there any more changes that you recommend ?

**Comment by jfrerich** [comment]
@maisnamrajusingh, I think the PR has to add the exclude option at the repo level.  Can you add that to the PR?

**Comment by maisnamraju** [comment]
> @maisnamrajusingh, I think the PR has to add the exclude option at the repo level. Can you add that to the PR?

@jfrerich  I think it is already covered in this PR. Let me know if you meant to do it at the organizational level ? in which case its not useful to the user. But let me know and @sibasankar-demansol  can add that to the PR.

**Comment by jasonblais** [comment]
cc @jfrerich @hanzei kind reminder to help with review

**Comment by jfrerich** [comment]
@maisnamraju, it took a bit for me to set up, but I wanted to test this PR. I just wanted to double-check the use case I was not understanding.

Here is what I did and I'm documenting here in case QA wants is interested or if they want to approve the PR based on my DEV QA results

* Create an org (https://github.com/jasonsorg)
* Create a repo in the org (https://github.com/jasonsorg/jasonsorg-junk-repo)
* Go through the Github plugin install process and set up instructions from the README.
* I made the org and repo public so that I didn't have permissions problems.
* `/github subscriptions add jasonsorg`
  * subscribe all to the org
* create an issue in the repo
  * verify notification is received in the channel
* `/github subscriptions add jasonsorg --exclude jasonsorg/jasonsorg-junk-repo`
  * subscribe all to the org but omit the `jasonsorg-junk-repo`
* create another issue in the repo
  * verify notification is NOT sent to the channel

`/github subscribe list` results.
image

**Comment by sibasankar-demansol** [comment]
> @maisnamraju, it took a bit for me to set up, but I wanted to test this PR. I just wanted to double-check the use case I was not understanding.
> 
> Here is what I did and I'm documenting here in case QA wants is interested or if they want to approve the PR based on my DEV QA results
> 
> * Create an org (https://github.com/jasonsorg)
> * Create a repo in the org (https://github.com/jasonsorg/jasonsorg-junk-repo)
> * Go through the Github plugin install process and set up instructions from the README.
> * I made the org and repo public so that I didn't have permissions problems.
> * `/github subscriptions add jasonsorg`
>   
>   * subscribe all to the org
> * create an issue in the repo
>   
>   * verify notification is received in the channel
> * `/github subscriptions add jasonsorg --exclude jasonsorg/jasonsorg-junk-repo`
>   
>   * subscribe all to the org but omit the `jasonsorg-junk-repo`
> * create another issue in the repo
>   
>   * verify notification is NOT sent to the channel
> 
> `/github subscribe list` results.
> image
> 
> To remove the subscription exclusion
> `/github subscriptions delete jasonsorg/jasonsorg-junk-repo`

@jfrerich  yes that's how it should work

File: server/plugin/command.go [hunk 21]
```diff
@@ -314,7 +325,30 @@ func (p *Plugin) handleSubscribesAdd(_ *plugin.Context, args *model.CommandArgs,
...
+				if err := p.StoreExcludedNotificationRepo(val); err != nil {
+					return err.Error()
+				}
+				if excludeMsg != "" {
+					excludeMsg += " and " + NotificationOffRepo
+					continue
+				}
+				excludeMsg += NotificationOffRepo
+			}
+			subOrgMsg += "\n\n" + fmt.Sprintf("But Notifications are disabled for %s", excludeMsg)
```
Comment (jfrerich) [inline server/plugin/command.go:346]: I think the messaging is good, but would remove the `But`.image
```suggestion
			subOrgMsg += "\n\n" + fmt.Sprintf("Notifications are disabled for %s", excludeMsg)
```

Comment (jfrerich) [reply server/plugin/command.go:346]: This is outside the scope of this ticket, but we could hyperlink the organization and repos :)

**Review by jfrerich — CHANGES_REQUESTED** [review]
I only have 1 more nitpick

File: server/plugin/command.go (hunk 21 above)
Comment (sibasankar-demansol) [reply server/plugin/command.go]: Sure , have added this too :)

File: server/plugin/subscriptions.go [hunk 22]
```diff
@@ -255,6 +273,66 @@ func (p *Plugin) StoreSubscriptions(s *Subscriptions) error {
 	return nil
 }
 
+func (p *Plugin) GetExcludedNotificationRepos() ([]string, error) {
+	var subscriptions []string
+	value, appErr := p.API.KVGet(SubscribedRepoNotificationOff)
```
Comment (hanzei) [inline server/plugin/subscriptions.go:278]: Is there one global list of `SubscribedRepoNotificationOff`? What happens if there are multiple  orgs with disabled repos?

**Review by hanzei — CHANGES_REQUESTED** [review]
One question

File: server/plugin/subscriptions.go (hunk 22 above)
Comment (sibasankar-demansol) [reply server/plugin/subscriptions.go:278]: @hanzei  he have taken care of this by storing the `org/repo` so that it won't have any issue when we have multiple orgs disabled

**Review by hanzei — APPROVED** [review]
Code looks good, but I have not tested the change

File: server/plugin/command.go (hunk 21 above)
Comment (jfrerich) [reply server/plugin/command.go]: Awesome! Thanks!

**Review by jfrerich — APPROVED** [review]
@sibasankarnayak, LGTM! Thanks for tackling this PR. I will be using it ASAP :)

**Comment by sibasankarnayak** [comment]
> @sibasankarnayak, LGTM! Thanks for tackling this PR. I will be using it ASAP :)

will be more happy to see getting used.

**Comment by hanzei** [comment]
Let's deffer QA to release testing.
